transformers4rec/torch/model/prediction_task.py

```diff
@@ -417,50 +438,249 @@ class _NextItemPredictionTask(torch.nn.Module):
+            logits = inputs @ output_weights.t()
 
         if self.softmax_temperature:
             # Softmax temperature to reduce model overconfidence
             # and better calibrate probs and accuracy
             logits = torch.div(logits, self.softmax_temperature)
 
-        predictions = self.log_softmax(logits)
+        return logits, targets
+
+    def sampled(self, inputs, targets, output_weights):
+        """Returns logits using sampled softmax"""
+        neg_samples, targets_probs, samples_probs = self.sampler.sample(targets)
 
-        return predictions
+        positive_weights = output_weights[targets]
+        negative_weights = output_weights[neg_samples]
+
+        positive_scores = (inputs * positive_weights).sum(dim=-1, keepdim=True)
+        negative_scores = inputs @ negative_weights.t()
+
+        # logQ correction, to not overpenalize popular items for being sampled
+        # more often as negatives
+        epsilon = 1e-16
+        positive_scores -= torch.unsqueeze(torch.log(targets_probs + epsilon), dim=-1)
+        negative_scores -= torch.unsqueeze(torch.log(samples_probs + epsilon), dim=0)
+
+        # Remove accidental matches
+        accidental_hits = torch.unsqueeze(targets, -1) == torch.unsqueeze(neg_samples, 0)
+        negative_scores[accidental_hits] = torch.finfo(torch.float16).min / 100.0
+
+        logits = torch.cat([positive_scores, negative_scores], axis=1)
+        new_targets = torch.zeros(logits.shape[0], dtype=torch.int64, device=targets.device)
+
+        return logits, new_targets
 
     def _get_name(self) -> str:
         return "NextItemPredictionTask"
+
+
+class LogUniformSampler:
+    def __init__(
+        self,
+        max_n_samples: int,
+        max_id: int,
+        min_id: Optional[int] = 0,
+        unique_sampling: bool = True,
+        n_samples_multiplier_before_unique: int = 2,
+    ):
+        """LogUniformSampler samples negative samples based on a log-uniform distribution.
+        `P(class) = (log(class + 2) - log(class + 1)) / log(max_id + 1)`
+
+        This implementation is based on to:
+        https://github.com/kimiyoung/transformer-xl/blob/master/pytorch/utils/log_uniform_sampler.py
+        TensorFlow Reference:
+        https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/ops/candidate_sampling_ops.py
+
+        LogUniformSampler assumes item ids are sorted decreasingly by their frequency.
+
+        if `unique_sampling==True`, then only unique sampled items will be returned.
+        The actual # samples will vary from run to run if `unique_sampling==True`,
+        as sampling without replacement (`torch.multinomial(..., replacement=False)`) is slow,
+        so we use `torch.multinomial(..., replacement=True).unique()` which doesn't guarantee
+        the same number of unique sampled items. You can try to increase
+        n_samples_multiplier_before_unique to increase the chances to have more
+        unique samples in that case.
```

Comment:
+1 !! Thank you for creating this class. It was very helpful for learning how to approximate item frequency distributions for both sampling with and without repetition!